yt/chyt/server/job_size_constraints.cpp

```diff
@@ -49,11 +59,78 @@ IJobSizeConstraintsPtr CreateClickHouseJobSizeConstraints(
     }
 
     auto inputSliceDataWeight = std::max<i64>(1, dataWeightPerJob * 0.1);
-    if (inputSliceDataWeight < config->MinSliceDataWeight) {
-        inputSliceDataWeight = dataWeightPerJob;
+
+    TJobSizeTrackerOptions jobSizeTrackerOptions;
+
+    if (readInOrderMode != EReadInOrderMode::None) {
+        jobSizeTrackerOptions = TJobSizeTrackerOptions{
+            .LimitProgressionRatio = 2,
+            .GeometricResources = {EResourceKind::DataWeight, EResourceKind::PrimaryDataWeight},
+            .LimitProgressionLength = 1,
+        };
+
+        // This allows configuring the size of the first subquery via query settings.
+        // We can logically use this setting, since each thread subquery forms its own
+        // secondary query when reading in order.
+        auto startDataWeightPerJob = std::max(executionSettings->MinDataWeightPerSecondaryQuery, subqueryConfig->MinDataWeightPerThread);
+        for (auto dataWeight = startDataWeightPerJob; dataWeight < dataWeightPerJob; dataWeight *= *jobSizeTrackerOptions.LimitProgressionRatio) {
+            ++jobSizeTrackerOptions.LimitProgressionLength;
+        }
+
+        switch (readInOrderMode) {
+            case EReadInOrderMode::Forward:
+                // The first range is small!
+                dataWeightPerJob = startDataWeightPerJob;
+                break;
+            case EReadInOrderMode::Backward:
+                // We will be reading these ranges backwards!
+                jobSizeTrackerOptions.LimitProgressionRatio = 0.5;
+                // We need to perform an actual split for the optimization to work.
+                if (jobCount == 1) {
```

Comment:
No, because in forward mode we start with `startDataWeightPerJob`, which is small, instead of `dataWeightPerJob` computed above. If this small size is larger than the entire table, then that's fine and we only need one small job. 

The problem in backward mode is that we need to produce a few slices first before starting our exponential descent, and here we have no chance of doing that if we put everything in one single slice in the very beginning.